feature: add a refactored copy of botticelli

bassosimone opened the pull request:
This diff is a quick refactor of [neubot/botticelli](
https://github.com/neubot/botticelli) that compiles and
will hopefully work. I started moving around all the
pieces that were part of botticelli to ensure that the
overall code is more tidy.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/ndt-server-go/15)
<!-- Reviewable:end -->

bassosimone committed:
feature: add a refactored copy of botticelli

This diff is a quick refactor of [neubot/botticelli](
https://github.com/neubot/botticelli) that compiles and
will hopefully work. I started moving around all the
pieces that were part of botticelli to ensure that the
overall code is more tidy.

gfr10598 commented:
*[ndt-server.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yp_NpGHNP3sJEk-VX:-L3Yp_NpGHNP3sJEk-VY:br8151a) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/ndt-server.go#L1)):*
> ```Go
> package main
> ```

This should go under cmd/ndt-server/ndt-server.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[protocol/messages.go, line 26 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YrhoUgKq6Z94iWCx_:-L3YrhoUgKq6Z94iWCxa:b-l5tghj) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/protocol/messages.go#L26)):*
> ```Go
> */
> 
> func read_message_internal(cc net.Conn, reader io.Reader) (
> ```

readMessageInternal

---

*[util/io.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-L3Yq6yH3UBP4VK5eVsn:b3b43xa) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
> ```Go
> // Part of Neubot <https://neubot.nexacenter.org/>.
> ```

Let's pull util into a separate PR, and add unit tests.

---

*[util/io.go, line 17 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YqwoOyelUxYRN2Yco:-L3YqwoOyelUxYRN2Ycp:b-fhffdy) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L17)):*
> ```Go
> 	"time"
> )
> ```

type DeadlineReader struct {
   io.Reader   // this inherits all methods of reader.  We will override them
   conn net.Conn  // This encapsulates the conn for deadlines.
   deadline time.Duration
}

func NewDeadlineReader(conn net.Conn) {
   // initialize deadline constant
   // create reader
   // create and return new struct
}

func (conn DeadlineReader) ReadFull(data) error {
   ...
}

---

*[util/io.go, line 30 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YqT4H8bjdLVDzzw8c:-L3YqT4H8bjdLVDzzw8d:bbe0mwr) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L30)):*
> ```Go
> 		return count, err
> 	}
> 	err = conn.SetReadDeadline(time.Time{})
> ```

return count, conn.SetReadDeadline(time.Time{})

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/io.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-L3YtjXa3H08pwZ5CXeF:bm3hwtz) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Let's pull util into a separate PR, and add unit tests.
</blockquote></details>

Actually, let's start with just util/random first.

---

*[util/io.go, line 30 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3YqT4H8bjdLVDzzw8c:-L3YtRz1GdfJG9FlH-i4:b-f0ou9s) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L30)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

return count, conn.SetReadDeadline(time.Time{})
</blockquote></details>

(If we keep this outline)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[util/io.go, line 1 at r1](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-L3Yu4IU4-GkK87lkIdQ:batuklf) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Actually, let's start with just util/random first.
</blockquote></details>

For this one, maybe use package ioutil

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

bassosimone commented:
Started moving the code over in #18

bassosimone committed:
Merge branch 'master' into feature/botticelli

Command that I run: `git merge -X theirs master`

bassosimone committed:
fix(s2c.go): use util.BytesGenerator

bassosimone committed:
fix(botticelli): use part of I/O code

bassosimone committed:
cleanup(u/io.go): remove unneeded code

Since we're now using code in netx, we can clean up code
that we don't use from the utils/io.go file.

bassosimone committed:
messages.go: add TODOs for me and @gfr10598

bassosimone commented on protocol/constants.go:
```diff
@@ -0,0 +1,34 @@
+
+// Message types:
+
+const KvCommFailure byte = 0
```
TODO(bassosimone): use `iota` to initialise variables

bassosimone commented on protocol/constants.go:
```diff
@@ -0,0 +1,34 @@
+package protocol
```
This is part of the protocol

bassosimone replied:
It can also be at the top of one file

bassosimone commented on protocol/constants.go:
```diff
@@ -0,0 +1,34 @@
+const KvSrvQueue byte = 1
+const KvMsgLogin byte = 2
+const KvTestPrepare byte = 3
+const KvTestStart byte = 4
```
Drop the `kv` prefix

bassosimone commented on protocol/control.go:
```diff
@@ -0,0 +1,178 @@
+			return
+		}
+		time.Sleep(3.0 * time.Second)
+	}
```
TODO: remove the queue before submitting PR since it's suboptimal

bassosimone commented on protocol/control.go:
```diff
@@ -0,0 +1,178 @@
+	// possibly also very ugly and stupid. Must be improved.
+	//
+	// Moreover the lock/unlock dance with the mutex is not idiomatic
+	// golang and it would be better to use messages and channels.
```
TODO: separate before the queue and after the queue

bassosimone commented on protocol/meta.go:
```diff
@@ -0,0 +1,52 @@
+package protocol
+
```
single package for all the tests; let's use `nettests`

bassosimone commented on protocol/s2c.go:
```diff
@@ -0,0 +1,169 @@
+	// Bind port and tell the port number to the server
+	// TODO: choose a random port instead than an hardcoded port
+
+	listener, err := net.Listen("tcp", ":3010")
```
Let's take code for random port from the "start" branch!

bassosimone commented on protocol/s2c.go:
```diff
@@ -0,0 +1,169 @@
+
+*/
+
+const kv_parallel_streams int = 3
```
Add reference to the original spec

bassosimone committed:
Refactor way in which we listen with deadline

1. rather than setting the deadline every time we attempt to
   Accept() just set it once when constructing the TCP listener

2. Update surrounding code

3. While there, plug possible listener leak

bassosimone committed:
Merge branch 'master' into feature/botticelli

bassosimone committed:
Fix merge

bassosimone committed:
protocol.go: import botticelli constants

bassosimone committed:
protocol.go: import botticelli constants

bassosimone committed:
Update botticelli code to renamed constants

bassosimone committed:
fix typo

bassosimone committed:
protoocol.go: remove TestCode

The botticelli code assumes that TestCode is an int. I think we can
safely keep that assumption. Otherwise, if Greg does not like this
way of proceeding, we can revert this diff and add casts.

bassosimone committed:
refactor(protocol.go): sniff for WebSockets during Login

I would argue that we don't need to sniff for WebSockets at every
message, rather we want to do that only when reading Login.

In fact, as far as I understand, we should pick a protocol at the
beginning, and then stick to it for the rest of the session.

As such, it only matters to sniff for WebSockets during Login.

Added also a TODO comment explaining how I am planning to go about
further specializing the code next, so Greg can comment on it.

bassosimone committed:
Address feedback by @gfr10598

- remove excess whitespaces

- remove code for sniffing web-sockets, as we will do that
  on a separate handler, for the sake of simplicity

bassosimone committed:
Zap unused variable

bassosimone committed:
Update .gitignore

This small change can sneak in as part of the PR.

bassosimone committed:
Revert "refactor(protocol.go): sniff for WebSockets during Login"

This reverts commit 9e7c869dcea7c621385efccf2ae34c41129697cc.

bassosimone committed:
Revert "protoocol.go: remove TestCode"

This reverts commit 4e64efa644700730c31f5510cbd6bb6773ebf2cc.

bassosimone committed:
Merge remote-tracking branch 'origin/refactor/protocol_codes' into feature/botticelli

bassosimone committed:
Use TestCode type for type-safety

See #25

bassosimone committed:
messages.go: start refactoring read_message_internal

This diff contains a number of questions for @gfr10598.

bassosimone committed:
messages.go: move JSON reading code to protocol.go

bassosimone committed:
messages.go: use protocol.go to read login message

bassosimone committed:
Merge branch 'master' into feature/botticelli

bassosimone committed:
Get rid of ReadJsonMessage

bassosimone committed:
Get rid of WriteRaWstring

LOL for the capitalization, I just noticed it now.

bassosimone committed:
Get rid of ReadExtendedLogin

bassosimone committed:
Finish getting rid of protocol/messages.go

bassosimone committed:
Remove commented out code

bassosimone committed:
Separate packages

bassosimone committed:
Get rid of the queue

bassosimone committed:
Extra style fixes

gfr10598 commented:
Review status: 0 of 5 files reviewed at latest revision, 13 unresolved discussions.

---

*[nettests/meta.go, line 6 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85CJbC4dCDOgH5Hbc:-LD85CJbC4dCDOgH5Hbd:b-qa067a) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L6)):*
> ```Go
> 	"bufio"
> 	"errors"
>   "github.com/m-lab/ndt-server-go/protocol"
> ```

go fmt

---

*[nettests/meta.go, line 26 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85pEr0_GJsdnGR1Bl:-LD85pEr0_GJsdnGR1Bm:bb69l0f) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L26)):*
> ```Go
> 
> 	// Read a sequence of TEST_MSGs from client
> ```

extra newline

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

bassosimone committed:
Repair regress

bassosimone committed:
s2c.go: use a random port

bassosimone committed:
Miscellaneous style fixes

bassosimone committed:
Move ndt-server command in subdirectory

bassosimone commented:
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions.

---

*[protocol/constants.go, line 1 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9I-r2-1:-LD8S-N7Dbl0OQeEdEK9:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/constants.go#L1)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

It can also be at the top of one file
</blockquote></details>

Done.

---

*[protocol/constants.go, line 5 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9I-r2-5:-LD8RzS0E3ykmamKzwxj:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/constants.go#L5)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO(bassosimone): use `iota` to initialise variables
</blockquote></details>

Done.

---

*[protocol/constants.go, line 9 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9I-r2-9:-LD8S2_E7n80DHHxf3RJ:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/constants.go#L9)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Drop the `kv` prefix
</blockquote></details>

Done.

---

*[protocol/control.go, line 70 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9J-r2-70:-LD8S4js0pK07Ot6nKV5:bmg78ov) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/control.go#L70)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO: separate before the queue and after the queue
</blockquote></details>

Not relevant since we've decided not to have a queue

---

*[protocol/control.go, line 86 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9J-r2-86:-LD8S3ueBwjh1rEwyW29:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/control.go#L86)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

TODO: remove the queue before submitting PR since it's suboptimal
</blockquote></details>

Done.

---

*[protocol/messages.go, line 26 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YrhoUgKq6Z94iWCx_:-LD8SCzmA7R6ohD_XNac:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/protocol/messages.go#L26)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

readMessageInternal
</blockquote></details>

Done.

---

*[protocol/meta.go, line 2 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9L-r2-2:-LD8R-rU2I4IdGO9igBy:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/meta.go#L2)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

single package for all the tests; let's use `nettests`
</blockquote></details>

Done.

---

*[protocol/s2c.go, line 24 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9M-r2-24:-LD8ROsCCd5rqaAewriZ:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/s2c.go#L24)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Add reference to the original spec
</blockquote></details>

Done.

---

*[protocol/s2c.go, line 40 at r2](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YpB9d4mlAB87sTL9M-r2-40:-LD8RrUD7LtPl0dudcZl:b3m4z5u) ([raw file](https://github.com/m-lab/ndt-server-go/blob/e916e1b318eb754d27bc56c9d9f191ba7c127afb/protocol/s2c.go#L40)):*
<details><summary><i>Previously, bassosimone (Simone Basso) wrote…</i></summary><blockquote>

Let's take code for random port from the "start" branch!
</blockquote></details>

Used more or less equivalent code.

---

*[util/io.go, line 1 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3Yq6yGkJ1ES2fMj9uq:-LD8SH5G5dHXNzn5PHqK:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

For this one, maybe use package ioutil
</blockquote></details>

Done.

---

*[util/io.go, line 17 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YqwoOyelUxYRN2Yco:-LD8SJVC8NUw6_OKlQo4:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L17)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

type DeadlineReader struct {
   io.Reader   // this inherits all methods of reader.  We will override them
   conn net.Conn  // This encapsulates the conn for deadlines.
   deadline time.Duration
}

func NewDeadlineReader(conn net.Conn) {
   // initialize deadline constant
   // create reader
   // create and return new struct
}

func (conn DeadlineReader) ReadFull(data) error {
   ...
}
</blockquote></details>

Done.

---

*[util/io.go, line 30 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3YqT4H8bjdLVDzzw8c:-LD8SIXq858XuWSuOc_4:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/util/io.go#L30)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

(If we keep this outline)
</blockquote></details>

Done.

---

*[nettests/meta.go, line 6 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85CJbC4dCDOgH5Hbc:-LD8R9Nd1cqGGKBnfdgq:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L6)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

go fmt
</blockquote></details>

Done.

---

*[nettests/meta.go, line 26 at r3](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-LD85pEr0_GJsdnGR1Bl:-LD8RHeUBEoJ_rkKYyPT:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b08e404db670a2413fe19646f115bf96a70ed8b5/nettests/meta.go#L26)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

extra newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

bassosimone commented:
Review status: 0 of 5 files reviewed at latest revision, 15 unresolved discussions.

---

*[ndt-server.go, line 1 at r1](https://reviewable.io/reviews/m-lab/ndt-server-go/15#-L3Yp_NpGHNP3sJEk-VX:-LD8SO756nRt-euJ_Eoy:b-896fix) ([raw file](https://github.com/m-lab/ndt-server-go/blob/b0f9f86685337db1ed617e48588b1601c8dfdeca/ndt-server.go#L1)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This should go under cmd/ndt-server/ndt-server.go
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/m-lab/ndt-server-go/15)*
<!-- Sent from Reviewable.io -->

bassosimone committed:
Fix queue management

bassosimone committed:
Fix gitignore

bassosimone committed:
Unwrap long line

bassosimone committed:
Remove comments that I now have addressed

bassosimone commented on control/control.go:
```diff
@@ -0,0 +1,145 @@
+		return
+	}
+
+	// Queue management (simplified)
```
May want to call this dispatcher rather than queue

pboothe commented:
Closing this.
